javascript/pkce_oauth/server/index.js

```diff
@@ -0,0 +1,174 @@
+
+    let state = req.query.state;
+
+    // Confirm the given state matches the state in the user's cookies
+    if (state !== req.cookies.state) {
+        res.status(401);
+        res.send();
+        return;
+    }
+
+    // The user doesn't need this state anymore
+    res.clearCookie("state");
+
+    let code = req.query.code;
+    let codeVerifier = stateCache[state];
+
+    const requestBody = {
+        grant_type: authorizationGrantType,
+        client_id: clientId,
+        client_secret: clientSecret,
+        code_verifier: codeVerifier,
+        redirect_uri: redirectUri,
+        code: code,
+    };
+
+    // Lets ask the auth server for an access and refresh token
+    request.post(tokenExchangeEndpoint, {
+        form: requestBody
+    }, (error, responseObj, responseBody) => {
+        console.log('In response from asana');
+
+        if (error) {
+            console.error(error);
+            res.status(401);
+            res.send();
+            return;
+        }
+
+        handleNewToken(JSON.parse(responseBody), res);
+    });
+});
+
+// This allows the front end app to request a new token if their current token is not expired
+app.get('/refresh-token', (req, res) => {
+    console.log('In refresh-token');
+
+    let access_token = req.cookies.access_token;
+    let refreshTokenObj = tokenCache[access_token];
+
+    // If this token is not expired, refresh it
+    if (refreshTokenObj != null && Date.now() - refreshTokenObj.createdAt < refreshTokenObj.expiresIn) {
+        const requestBody = {
+            grant_type: refreshGrantType,
+            client_id: clientId,
+            client_secret: clientSecret,
+            refresh_token: refreshTokenObj.refreshToken,
+        };
+
+        // Ask the auth server for a new token
+        request.post(tokenExchangeEndpoint, {
+            form: requestBody
+        }, (error, responseObj, responseBody) => {
+            if (error) {
+                console.error(error);
+                return;
+            }
+
+            handleNewToken(JSON.parse(responseBody), res);
+        });
+    } else {
+        res.statusCode = '401';
+        res.send();
+    }
+});
+
+// Handles a response from the auth server.
+function handleNewToken(body, res) {
+    let accessToken = body.access_token;
+    let expiresIn = body.expires_in;
+    let refreshToken = body.refresh_token;
+    let userData = body.data;
+
+    console.log("Storing access & refresh token for " + userData.name + " (" + userData.gid + ")");
+
+    res.cookie("access_token", accessToken);
+
+    tokenCache[accessToken] = {
+        createdAt: Date.now(),
+        expiresIn: expiresIn,
+        refreshToken: refreshToken
+    };
+
+    res.redirect(redirectUri);
+}
+
+// Generates a random string of a bigger size
+function getRandomToken() {
+    return getRandomSmallToken() + getRandomSmallToken();
+}
+
+// Generates a random string
+function getRandomSmallToken() {
+    let randomNumber = Math.random();
```

Comment:
Use a cryptographically random value please! See [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) for an example.

Reply:
Looks like I don't have access to crypto in nodejs, but they have an equivalent.